From a59cda6c1df53ae361ff4ed6cdce3bec10f36fd1 Mon Sep 17 00:00:00 2001
From: Marek Kasik <mkasik@redhat.com>
Date: Wed, 31 Jul 2013 14:58:40 +0200
Subject: [PATCH 06/12] print-notifications: Cancel subscription to DBus
 signals when finished

Store id of the subscription to CUPS' DBus signals and remove it
when stopping manager.
---
 .../gsd-print-notifications-manager.c              | 29 ++++++++++++++--------
 1 file changed, 19 insertions(+), 10 deletions(-)

diff --git a/plugins/print-notifications/gsd-print-notifications-manager.c b/plugins/print-notifications/gsd-print-notifications-manager.c
index 379d67f..f7a8463 100644
--- a/plugins/print-notifications/gsd-print-notifications-manager.c
+++ b/plugins/print-notifications/gsd-print-notifications-manager.c
@@ -75,6 +75,7 @@ struct GsdPrintNotificationsManagerPrivate
         GHashTable                   *printing_printers;
         GList                        *active_notifications;
         guint                         cups_connection_timeout_id;
+        guint                         cups_dbus_subscription_id;
 };
 
 enum {
@@ -1069,16 +1070,17 @@ gsd_print_notifications_manager_got_dbus_connection (GObject      *source_object
         manager->priv->cups_bus_connection = g_bus_get_finish (res, &error);
 
         if (manager->priv->cups_bus_connection != NULL) {
-                g_dbus_connection_signal_subscribe (manager->priv->cups_bus_connection,
-                                                    NULL,
-                                                    CUPS_DBUS_INTERFACE,
-                                                    NULL,
-                                                    CUPS_DBUS_PATH,
-                                                    NULL,
-                                                    0,
-                                                    on_cups_notification,
-                                                    manager,
-                                                    NULL);
+                manager->priv->cups_dbus_subscription_id =
+                        g_dbus_connection_signal_subscribe (manager->priv->cups_bus_connection,
+                                                            NULL,
+                                                            CUPS_DBUS_INTERFACE,
+                                                            NULL,
+                                                            CUPS_DBUS_PATH,
+                                                            NULL,
+                                                            0,
+                                                            on_cups_notification,
+                                                            manager,
+                                                            NULL);
         } else {
                 g_warning ("Connection to message bus failed: %s", error->message);
                 g_error_free (error);
@@ -1146,6 +1148,13 @@ gsd_print_notifications_manager_stop (GsdPrintNotificationsManager *manager)
         manager->priv->num_dests = 0;
         manager->priv->dests = NULL;
 
+        if (manager->priv->cups_dbus_subscription_id > 0 &&
+            manager->priv->cups_bus_connection != NULL) {
+                g_dbus_connection_signal_unsubscribe (manager->priv->cups_bus_connection,
+                                                      manager->priv->cups_dbus_subscription_id);
+                manager->priv->cups_dbus_subscription_id = 0;
+        }
+
         if (manager->priv->subscription_id >= 0)
                 cancel_subscription (manager->priv->subscription_id);
 
-- 
1.8.3.1

